source/common/router/upstream_request.cc

```diff
@@ -398,6 +412,15 @@ void UpstreamRequest::onPoolReady(
   }
 }
 
+void UpstreamRequest::onStreamMaxDurationReached() {
+  ASSERT(upstream_host_);
+  upstream_host_->cluster().stats().upstream_rq_max_duration_reached_.inc();
+  max_stream_duration_timer_->disableTimer();
+  max_stream_duration_timer_.reset();
+  resetStream();
+  parent_.callbacks()->resetStream();
```

Comment:
It seems a little odd to me that we have to explicitly invoke the reset callbacks here after we reset the stream in the previous line. cc @alyssawilk to take a look at this given the recent factor. I would have assumed that this would be encapsulated somehow in the reset getting passed through to the core router. If we do need this here we should have some more comments.

Reply:
If we didn't do this in this line. I think that downstream never reset. So we need this, don't we?

Reply:
> So we need this, don't we?

I don't know, it's not intuitive from looking at the code. So it either needs more comments or needs to be refactored slightly to make it more clear. I will defer to @alyssawilk on this part.

Reply:
I guess I'm wondering if we really want to reset the stream here at all.

It seems like in the case of hedging or retriable requests, what we want to do is close the upstream connection and let the router hedge / retry logic sort out if the stream can continue, or if it should send a local reply due to the (locally caused) disconnect.

Reply:
yep. This is to say if it is a retriable request, we should retry along with retry policy. If not, we should close downstream and send a local reply. ok?

Reply:
Yes I think this makes more sense. IMO this timeout should work the same as the per-try-timeout. Just let the parent of the upstream request figure out what to do.